src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1083,8 +1098,33 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreamsWithOrder(
             splitted_parts_and_ranges.emplace_back(std::move(new_parts));
         }
 
+        bool primary_key_type_supports_virtual_row = true;
+        const auto & actions = storage_snapshot->metadata->getPrimaryKey().expression->getActions();
+        for (const auto & action : actions)
+        {
+            if (action.node->type != ActionsDAG::ActionType::INPUT)
+            {
+                primary_key_type_supports_virtual_row = false;
+                break;
+            }
+        }
+
+        /// If enabled in the optimization stage, check whether there are more than one branch.
```

Comment:
I suppose we can allow virtual rows for any number of parts. It should be relatively cheap.